File: Matlab/Algorithms/binarySearch.m
function [imid] =  binarySearch(array, value, imin, imax, recursion)
%%  This function implements binary search on an array given the value of interest

    if ~recursion 
        %Iterative method
        flagFound = 0;
        while imax >= imin
            imid = floor((imax + imin)/2);
            if (array(imid) == value)
                fprintf('Value has index: %d. \n', imid);
                flagFound = 1;
                break
            elseif (array(imid) < value)
                fprintf('Value is located in right half, midpoint value = %d. \n', array(imid));
                imin = imid + 1;
            elseif (array(imid) > value)
                fprintf('Value is located in left half, midpoint value = %d. \n', array(imid));
                imax = imid - 1;
            end
        end
        if ~flagFound
            fprintf('Value not in array.\n');
        end
    elseif recursion  
        %Recursive method
        imid = floor((imax + imin)/2);
        if imin > imax
            fprintf('Value not in array');
        else
            if (array(imid) == value)
                fprintf('Value has index: %d. \n', imid);
            elseif (array(imid) < value)
                fprintf('Value is located in right half, midpoint value = %d. \n', array(imid));
                imid = binarySearch(array, value, imid+1, imax, recursion);
            elseif (array(imid) > value)
                fprintf('Value is located in left half, midpoint value = %d. \n', array(imid));
                imid = binarySearch(array, value, imin, imid-1, recursion);
            end
        end
    end
end

File: C/Algorithms/binarySearch.c

#include <stdio.h>
#include <math.h> 

int binarySearch(int numarray[],int value, int imin, int imax);


int binarySearch(int numarray[], int value, int imin, int imax)
{
int imid;
while (imin <=imax)
{
    imid = (int)(floor((double)(imin) + (double)(imax))/2);
    if (numarray[imid] == value)
    {
        return imid;
    }
    else if (numarray[imid] < value)
    {
        imin = imid + 1;
    }
    else if (numarray[imid] > value)
    {
        imax = imid - 1;
    }
}
    printf("Value not in array. \n");
    return -1;
}
 
/* This program runs binarySearch */ 
int  main(void)
{
    int index;
    int value = 7;
    int numarray[10] = {1,2,3,4,5,6,7,8,9,10};  	 
    index = binarySearch(numarray, value, 1, 10); 
    if (index >=0) 
    {
    	printf("Value %5d is at index %5d \n", numarray[index], index);
    } 
   return 0;
}
